### ISSUE TITLE
[Feature request] Automatic backup of savedata

### ISSUE THREAD

**Issue by rafael-57** [shadps4-emu/shadPS4#1909]
This would be very useful to prevent from savedata corruption (for example due to emulator crashes), game glitches, and just as a nice feature to have.

I first thought of copying the savedata into the backup folder once every few minutes, but this could still have you lose saves if you crash while this happens. 

I don't know how to code it but I would do it this way:

1. First, when the game asks to, write savedata to the usual destination
2. Every other time the game writes savedata, also write it in the automatic backup destination

What do you guys think?

**Comment by mailwl** [comment]
don't understand how it will work? if game crashed while save, it can't make backup anyway

**Comment by rafael-57** [comment]
> don't understand how it will work? if game crashed while save, it can't make backup anyway

Sorry, bad wording on my end. I meant something like this:
1. Write to savedata
2. Write savedata + backup
3. Write to savedata
4. Write savedata + backup etc.

This way at least we have an older copy

**Comment by GHU7924** [comment]
Doubtful, but OK.
I don't understand how it will work either. Is it possible to make a backup immediately after your save has occurred?
Is there a chance that your backup will happen during the save and you will receive an old broken copy of the save?
I can't tell you when the best time for backup will be.

In general, the idea of a backup save is pretty good, but the implementation of what it should look like is not yet clear.
This can help avoid a situation where you deleted the emulator, and then remembered that there were saves in the users section and you forgot to move them.

The first thing I can say is that it should be an unambiguously controlled decision.
I mean, the checkbox that the user decides to put or not, whether backup will be performed or not.

Secondly, this is the place where the backup will be stored (there are many options). Considering that we have the paths that we specify for the emulator, a copy of the saves can be stored there, that is, in the folder with the installed games (for example, the folder will be called SAVEDATA-BACKUP). Or maybe in the Documents folder. Or somewhere else.

**Comment by rafael-57** [comment]
> Doubtful, but OK. I don't understand how it will work either. Is it possible to make a backup immediately after your save has occurred? Is there a chance that your backup will happen during the save and you will receive an old broken copy of the save? I can't tell you when the best time for backup will be.
> 
> In general, the idea of a backup save is pretty good, but the implementation of what it should look like is not yet clear. This can help avoid a situation where you deleted the emulator, and then remembered that there were saves in the users section and you forgot to move them.
> 
> The first thing I can say is that it should be an unambiguously controlled decision. I mean, the checkbox that the user decides to put or not, whether backup will be performed or not.
> 
> Secondly, this is the place where the backup will be stored (there are many options). Considering that we have the paths that we specify for the emulator, a copy of the saves can be stored there, that is, in the folder with the installed games (for example, the folder will be called SAVEDATA-BACKUP). Or maybe in the Documents folder. Or somewhere else.

Yes, regardless of implementation I would keep one extra, older copy of the savedata.

I agree it's ok to have with a checkbox. I leave it to the devs to figure out the imlementation and the savedata location :P

**Comment by mailwl** [comment]
which game needs backup saves? maybe on exact game devs can fix savegame implementation

**Comment by rafael-57** [comment]
> which game needs backup saves? maybe on exact game devs can fix savegame implementation

No single game, this is a feature request for the entire emu.

**Comment by rainmakerv3** [comment]
I tried an implementation where it copies saves to a new folder after 5 minutes (can be adjusted and/or made configurable ofc). Seems to work, at least on my end, but there may or may not be a performance penalty caused by the new sleeping thread. 

Maybe some folks can test it: 
https://github.com/rainmakerv3/shadPS4/actions/runs/12525256861

Or maybe someone can improve the code:
https://github.com/rainmakerv3/shadPS4/commit/5a67a108839a0a8c3f6e0b8f313e519c5ebbb1f0

**Comment by GHU7924** [comment]
Actually, that's how I see the implementation.
![bs](https://github.com/user-attachments/assets/2012423f-bcb8-4f8b-9cea-258210766e67)

And it will be a controlled decision.

And in the "Paths" tab, you can add an analog Game folders, but only for saving. You can choose 1-2-3 paths where to store the saves. But I don't know if it's necessary to have more than one backup.

@rafael-57 I have proposed my own version of the vision, but since the proposal came from you, you need to think over the implementation.

**Comment by rafael-57** [comment]
> Actually, that's how I see the implementation. ![bs](https://github.com/user-attachments/assets/2012423f-bcb8-4f8b-9cea-258210766e67)
> 
> And it will be a controlled decision.
> 
> And in the "Paths" tab, you can add an analog Game folders, but only for saving. You can choose 1-2-3 paths where to store the saves. But I don't know if it's necessary to have more than one backup.
> 
> [@rafael-57](https://github.com/rafael-57) I have proposed my own version of the vision, but since the proposal came from you, you need to think over the implementation.

Manual backup looks great! But did you code it in a repository somewhere?

@viniciuslrangel since @rainmakerv3's PR was closed, did you figure out the issue with current automatic backups? Because one time I restored them and they were corrupted just as my current saves. Is it possible to keep an older copy of automatic backups since they are already being created?

**Comment by GHU7924** [comment]
>Manual backup looks great! But did you code it in a repository somewhere?

I don't know how to program, if that's what you're asking. I also don't know if the developers will allow this option. In the menu I mentioned earlier, it will most likely be inappropriate. I've tried several options and redrawn them several times. In the final version at the moment, it looked like this:

Before

![Image](https://github.com/user-attachments/assets/3370a4a0-9ee0-45f7-a56d-9a0bbd0fdbdc)

After

![Image](https://github.com/user-attachments/assets/d54e14d1-997d-43b7-8b96-10cdec5638df)

As you can see, you will be able to specify the path where your backup copy of the saves folder will be stored.
You can also send your saves to the backup folder with one button, and copy the saves from the backup to the folder of your current saves with the other.

Thanks to @DanielSvoboda , the ability to quickly access the desired tab has recently been added, which will simplify the task.

![Image](https://github.com/user-attachments/assets/9114d8b8-38eb-499f-9278-3ad529557fe1)

@rafael-57 In this case, you may not need automatic backups. The only question is whether anyone will take up the implementation of this idea.

I really understand you, because I went through Bloodborne back on emulator version 0.2.1, there was a lot of torment and a lot of backups, which even helped me a lot once or twice, because I once had my saves damaged in the late stages of the game, it was about 25 hours of passage, but thanks to the backup I managed to complete the whole game.

https://github.com/shadps4-emu/shadPS4/pull/1967#issuecomment-2565260489 I had such thoughts.
It would be great if someone would take up the implementation of my idea from this comment, but I can't demand it.

**Comment by GHU7924** [comment]
@rafael-57 @DanielSvoboda Or, if possible:

![Image](https://github.com/user-attachments/assets/dfb84ae8-fcef-43b1-bca0-f89286f74cf7)
![Image](https://github.com/user-attachments/assets/75c73d5a-a927-47ba-8154-6ab0f8fdb849)

These are just ideas drawn using Paint.

**Comment by GHU7924** [comment]
It so happened that the other day I was looking through mods for Bloodborne and saw this [mod](https://www.nexusmods.com/bloodborne/mods/307), which may be the perfect solution to your issue, although there is one thing, the keys used there are used by the emulator. If we build a similar tool into the emulator, there will be no automation, but you can make a backup at any time, which sounds good.
